File: tools/authoring/compare_fridge_orthographic_gates.py
"""Write per-view silhouette overlays and one combined metrics report."""
from pathlib import Path
import json
from PIL import Image

ROOT = Path(__file__).resolve().parents[2]
FOLDER = ROOT / "docs" / "reference-lock" / "vaccine-fridge-v1"
metrics = {}
for view in ("front", "side", "back"):
    reference = Image.open(FOLDER / f"{view}-mask-candidate.png").convert("L")
    candidate = Image.open(FOLDER / f"candidate-{view}.png").convert("RGBA").getchannel("A")
    overlay = Image.new("RGBA", reference.size, (0, 0, 0, 0))
    r, c, o = reference.load(), candidate.load(), overlay.load()
    intersection = union = reference_count = candidate_count = 0
    for y in range(reference.height):
        for x in range(reference.width):
            rv, cv = r[x, y] >= 128, c[x, y] >= 128
            reference_count += rv
            candidate_count += cv
            intersection += rv and cv
            union += rv or cv
            if rv and cv:
                o[x, y] = (255, 255, 255, 180)
            elif rv:
                o[x, y] = (255, 40, 40, 220)
            elif cv:
                o[x, y] = (40, 255, 80, 220)
    overlay.save(FOLDER / f"{view}-silhouette-overlay.png")
    metrics[view] = {
        "referencePixels": reference_count,
        "candidatePixels": candidate_count,
        "intersectionOverUnion": intersection / union if union else 0,
        "differentPixels": union - intersection,
    }
(FOLDER / "orthographic-gate-metrics.json").write_text(json.dumps(metrics, indent=2), encoding="utf-8")
print(json.dumps(metrics, indent=2))

File: test/compare-fridge-orthographic-gates.mjs
import { readFileSync, writeFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import path from 'node:path';
import { PNG } from 'pngjs';

const folder = fileURLToPath(new URL('../docs/reference-lock/vaccine-fridge-v1/', import.meta.url));
const metrics = {};

for (const view of ['front', 'side', 'back']) {
  const reference = PNG.sync.read(readFileSync(path.join(folder, `${view}-mask-candidate.png`)));
  const candidate = PNG.sync.read(readFileSync(path.join(folder, `candidate-${view}.png`)));
  if (reference.width !== candidate.width || reference.height !== candidate.height) {
    throw new Error(`${view} gate dimensions differ: reference ${reference.width}x${reference.height}, candidate ${candidate.width}x${candidate.height}`);
  }

  const overlay = new PNG({ width: reference.width, height: reference.height });
  let intersection = 0;
  let union = 0;
  let referencePixels = 0;
  let candidatePixels = 0;
  for (let pixel = 0; pixel < reference.width * reference.height; pixel++) {
    const offset = pixel * 4;
    const referenceVisible = reference.data[offset] >= 128;
    const candidateVisible = candidate.data[offset + 3] >= 128;
    if (referenceVisible) referencePixels++;
    if (candidateVisible) candidatePixels++;
    if (referenceVisible && candidateVisible) intersection++;
    if (referenceVisible || candidateVisible) union++;

    let rgba = [0, 0, 0, 0];
    if (referenceVisible && candidateVisible) rgba = [255, 255, 255, 180];
    else if (referenceVisible) rgba = [255, 40, 40, 220];
    else if (candidateVisible) rgba = [40, 255, 80, 220];
    overlay.data.set(rgba, offset);
  }
  writeFileSync(path.join(folder, `${view}-silhouette-overlay.png`), PNG.sync.write(overlay));
  metrics[view] = {
    referencePixels,
    candidatePixels,
    intersectionOverUnion: union ? intersection / union : 0,
    differentPixels: union - intersection,
  };
}

writeFileSync(path.join(folder, 'orthographic-gate-metrics.json'), `${JSON.stringify(metrics, null, 2)}\n`);
console.log(JSON.stringify(metrics, null, 2));

